Adding a navigational structure based on aip.dev

garrettjonesgoogle opened the pull request.

garrettjonesgoogle commented:
@suztomo @elharo at least 1 reviewer should pull this locally and run it in Jekyll, then poke around to make sure everything is in order

suztomo commented:
Checking out your branch...

suztomo commented:
From https://github.com/GoogleCloudPlatform/cloud-opensource-java/wiki/Tools#jekyll-for-github-pages:

```
# In docs directory
bundle install --path vendor/bundle
bundle exec jekyll serve
```

It worked.

![image](https://user-images.githubusercontent.com/28604/75382685-f36d5480-58a8-11ea-8882-6cf83d53fda7.png)

suztomo commented on docs/_includes/jlbp-header.html:
```diff
@@ -0,0 +1,96 @@
+        <nav class="glue-header__link-bar">
+          <ul id="list-1" class="glue-header__list">
+            <li class="glue-header__item{% if page.jlbp or page.jlbp_index %} glue-header--is-active{% endif %}">
+              <a class="glue-header__link" href="/" aria-level="1">Browse best practices</a>
```
I find this "Browse best practices" link redundant, because the logo "JLBPs" goes to the same top page.

garrettjonesgoogle replied:
It is technically redundant, but the explicit link helps people find what they are looking for easier - clicking on the logo might not be obvious. Also, I would argue that we shouldn't couple together the root of the site with the JLBP index - we might want to separate those ideas apart later. Lastly, if we add more on the header later, it will feel more "even" (look at the header of aip.dev).

suztomo replied:
Yes, that button may be useful in future, but not now. We can easily add it back to the site when needed.

garrettjonesgoogle replied:
@elharo what do you think?

garrettjonesgoogle replied:
I was thinking of an idea for headers to add. What about:

- Best practices (the existing one)
- Articles
- Reference
- Tools (new page - discuss the static linkage checker)

garrettjonesgoogle replied:
Also:

- Talks (e.g. https://saturnism.me/talk/surviving-dependency-hell-maven/)

elharo replied:
Good idea. Let's put that in a separate PR though.

garrettjonesgoogle replied:
Doing a separate PR seems fine for the header links I suggested here - but how about the single header link that is already present ("Browse best practices") which @suztomo wants to remove?

elharo replied:
The header bar and site search is the part of this PR I'm not sold on. The navigation and sidebar is a much more obvious win. I think we can move that forward quickly without having to vet all the JavaScript that's being added for Tipue.

garrettjonesgoogle replied:
I can remove the search functionality if that's a sticking point. However, there is one key problem with removing the whole header - it's what provides the link to navigate to the root of the site, and on mobile, it's the only way to navigate away from the current page (through the hamburger menu).

garrettjonesgoogle commented:
Why do you need to understand it? It's CSS... which I never understand...

garrettjonesgoogle commented:
The top bar has the new search functionality, which seems like a very useful feature.

At a high level, I tried to minimize the work to create this navigational structure by starting with aip.dev's structure and stripping out stuff we don't need. That process was pretty straightforward. Gutting the top bar would require more work, then we'd need even more work to put back a top bar later when we want one. 

Basically, I view the current structure in the PR as an MVP, and further changes are improvements on top of it.

saturnism commented:
my comnets are:
- remove the right-hand nav, seems redundant
- left-hand nav has Best Practices -> Best Practices Index, seems redundant as well.
- articles - I feel the current articles are actually intro info, as oppose to blogs.

I actually do like the site as-is, and agreed that it could use a backlink.

Proposed left-nav:

 * Introduction
  * Cause of Dependency Conflicts
  * Compiler and Dependency Conflicts
 * Best Practices
  * [JLBP-1] ...
  * ...
 * References
  * Glossary

garrettjonesgoogle commented:
@saturnism :

- For longer articles, the table of contents can be helpful to navigate to specific sections. This is standard on cloud.google.com - e.g. https://cloud.google.com/billing/docs/concepts
- I'm fine removing the "best practice index" link.
- I agree that the articles are not blogs (which is why I didn't use the term "blogs"). Do you think the term "articles" is misleading? I plan on writing additional articles, which will be less "introductory" the more I write.
- I personally like the idea of the intro articles at the top, but @elharo objected to putting them above the best practices in a prior PR of mine. Also, as I continue to write more articles, the list will increase in size. We may not want to hide the best practice list as the article list grows.

saturnism commented:
> @saturnism :
> 
> * For longer articles, the table of contents can be helpful to navigate to specific sections. This is standard on cloud.google.com - e.g. https://cloud.google.com/billing/docs/concepts

The standard on c.g.c (namely concepts) doesn't mean it's a best practice...

> * I'm fine removing the "best practice index" link.

+1 . The index may just be the home page itself as is today.

> * I agree that the articles are not blogs (which is why I didn't use the term "blogs"). Do you think the term "articles" is misleading? I plan on writing additional articles, which will be less "introductory" the more I write.

I do feel articles is misleading; otherwise, articles become a long list of dumping ground for any new content; grouping them logically structure is better, and we can be more focused on what content to add.

> * I personally like the idea of the intro articles at the top, but @elharo objected to putting them above the best practices in a prior PR of mine. Also, as I continue to write more articles, the list will increase in size. We may not want to hide the best practice list as the article list grows.

I can understand not having Introudction over-shadow the rules themselves.  I also dislike a lot of c.g.c article put "intro" or "concepts" at the end ;)

How do we group these articles together? I see them as introduction, or  just info on Dependency Conflicts.

* Best Practices
* More on Dependency Conflicts
  * What causes it
  * Why compiler cannot detect it
  * ... add more as we go.

garrettjonesgoogle commented:
> > @saturnism :
> > 
> > * For longer articles, the table of contents can be helpful to navigate to specific sections. This is standard on cloud.google.com - e.g. https://cloud.google.com/billing/docs/concepts
> 
> The standard on c.g.c (namely concepts) doesn't mean it's a best practice...

I'm not sure why best practices matter exactly. Consistency itself is a best practice anyway, right? What best practices are you trying to enforce here?

I'm not sure why you are referencing "concepts" in particular, besides the fact that it's the link I gave; you can visit any page under cloud.google.com and see the TOC on the right, e.g. https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types .

garrettjonesgoogle commented:
@saturnism :

> > * I personally like the idea of the intro articles at the top, but @elharo objected to putting them above the best practices in a prior PR of mine. Also, as I continue to write more articles, the list will increase in size. We may not want to hide the best practice list as the article list grows.
> 
> I can understand not having Introudction over-shadow the rules themselves. I also dislike a lot of c.g.c article put "intro" or "concepts" at the end ;)
> 
> How do we group these articles together? I see them as introduction, or just info on Dependency Conflicts.
> 
> * Best Practices
> * More on Dependency Conflicts
>   
>   * What causes it
>   * Why compiler cannot detect it
>   * ... add more as we go.

What do you think about me using "Concepts" instead of "Articles"? (I'm looking for a short name so that I can stick it on the header later)

Do you feel that the article titles should be in the sidebar, or is a link to a Concepts index page sufficient?

garrettjonesgoogle commented:
I have made some updates:

- Removed the search functionality
- Removed the "Browse best practices" link from the header
- Removed the "Best practices index" from the sidebar
- Renamed "Articles" to "Concepts"

PTAL

garrettjonesgoogle commented:
![Screenshot from 2020-02-28 16-11-05](https://user-images.githubusercontent.com/13341017/75596344-0ae14480-5a45-11ea-93f1-f8c3d1cc0726.png)

elharo commented on docs/_includes/jlbp-closing-scripts.html:
```diff
@@ -0,0 +1,9 @@
+<!-- Closing scripts -->
```
What does this script do?

garrettjonesgoogle replied:
@lukesneeringer could you help us understand what the scripts in https://github.com/googleapis/aip/blob/master/_includes/aip-closing-scripts.html do and why we might need them?

garrettjonesgoogle replied:
- The inline one makes the header bar work on mobile.
- The other two are what Glue tells you to use. https://glue-docs.appspot.com/get-started/cdn/

garrettjonesgoogle replied:
I will add comments to the file explaining this.

elharo commented on docs/_layouts/default.html:
```diff
@@ -1,43 +1,42 @@
-    </script>
-    {% endif %}
-  </head>
+<html lang="{{ site.lang | default: 'en-US' }}" class="glue-flexbox">
```
stet. This is en-US, not site.lang

garrettjonesgoogle replied:
I don't understand what you're asking for...

garrettjonesgoogle replied:
I will revert this line.

elharo commented on docs/_sass/callouts.scss:
```diff
@@ -0,0 +1,63 @@
+  border: 1px solid $glue-blue-200;
+}
+
+.tldr {
```
which of these do we actually use? I don't see .tldr referenced anywhere for example

garrettjonesgoogle replied:
We don't use them yet, but I thought they could be useful to incorporate.

garrettjonesgoogle replied:
I will remove the callout stuff since it's not used.

elharo commented on docs/_sass/colors.scss:
```diff
@@ -0,0 +1,139 @@
+// This is forked. Do not edit.
```
why not? It it's not editable it shouldn't be in the repo.

garrettjonesgoogle replied:
I'm not sure what your objection is... it's pretty standard practice to put warnings on files that are derived from other sources. Technically such files can be edited, but it's not recommended because it's much harder to maintain.

elharo commented on docs/assets/images/glue-icons.svg:
```diff
@@ -0,0 +1,42 @@
+<svg xmlns="http://www.w3.org/2000/svg">
```
Why are including glue icons?

garrettjonesgoogle replied:
They are standard icons that we should use if we have use cases that match them.

garrettjonesgoogle commented:
I pulled out the page metadata and dynamic index into https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1270 .

garrettjonesgoogle commented:
Update:

- I have merged https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1270 , shrinking the size of this PR a bit (in terms of files touched)
- I consolidated most of the scss files into one
- I stripped out most of the css that wasn't referenced

Hopefully that helps the review go more smoothly.

garrettjonesgoogle commented:
I have removed the glue.scss file by applying styles manually with javascript. There was already an include of the glue file from the cdn, so the site still looks the same.

garrettjonesgoogle commented:
PTAL - I think this is ready for another round; it is significantly reduced in size, while still maintaining the same navigation and look & feel.

suztomo commented on docs/_sass/colors.scss:
```diff
@@ -0,0 +1,143 @@
+// These variables don't appear in the generated Glue css, which is why
+//   this copy is needed.
+// Please don't edit this file since it would create a maintenance
+//   burden for people trying to keep it up to date with Glue.
```
"Do not edit this file" is a good message. Would you further add a comment on what should be done in case we want to update this file?

garrettjonesgoogle replied:
It's located on git on borg, so I can't reference the original source publicly. Do you have a preference on how to provide the reference to people who need it?

suztomo replied:
'Search for "glue project" in git-on-borg' or using "go" link would be nice.

garrettjonesgoogle replied:
I will add a go link.

elharo commented on docs/concepts-index.md:
```diff
@@ -0,0 +1,8 @@
+---
+permalink: /concepts-index
+---
+# Concepts index
```
delete "index" from the title

garrettjonesgoogle replied:
Since I'm including the list of concept pages in the sidebar, I don't need this anymore.

elharo commented on docs/_includes/jlbp-nav.html:
```diff
@@ -0,0 +1,20 @@
+    {% endif %}{% endfor -%}
+    <li class="nav-item nav-item-header">Concepts</li>
+    <li class="nav-item{% if page.url == '/concepts-index' %} nav-item-active{% endif %}">
+      <a href="/concepts-index">Concepts index</a>
```
this is a section with one link. Either make it a single link to Concepts, or list the concept pages individually.

garrettjonesgoogle replied:
I'll just list the concept pages here.

garrettjonesgoogle commented on docs/_layouts/default.html:
```diff
@@ -1,49 +1,42 @@
+      {{ content }}
+      {% else %}
+      <main role="main" id="page-content">
+        {% include jlbp-nav.html %}
```
Ok I'll move it up/out.

garrettjonesgoogle commented:
> Viewing source the left sidebar, at least, is inside `main` element. Per HTML 5 spec it should be outside:

This will be resolved by moving the jlbp-nav include outside of `<main>`.

garrettjonesgoogle commented:
> JLBP number is styled as an h4. This should be done with CSS and it should probably be part of the h1 header. In general, you don't want an h4 appearing before the h1.

I thought it was odd too.

The JLBP number in the h4 is provided by the layout (`default.html`) right above the content. Thus, the JLBP number and the title couldn't be part of the same h1. Any recommendations on how to approach this? Should the JLBP number be in an h1 with a special class & special css style?

garrettjonesgoogle commented:
> View Source and Edit This page links are much the same. Can we drop one of them? Probably View Source since that means something a little different in browser contexts.

How about renaming "View source" to "View page on GitHub"? (I'll remove the "Edit this page" link)

garrettjonesgoogle commented:
> Could the View on Github link go straight to the docs directory instead of the main page of the project?

Will do.

garrettjonesgoogle commented:
> The meta description is incorrect:
> 
> ```
> <meta name="description" content="Code health dashboard for GCP open source projects" />
> <meta property="og:description" content="Code health dashboard for GCP open source projects" />
> ```
> 
> In the exiting site it is
> 
> ```
> 
> <meta name="description" content="Tools for detecting and avoiding linkage errors in GCP open source projects" />
> <meta property="og:description" content="Tools for detecting and avoiding linkage errors in GCP open source projects" />
> ```
> 
> which is a little better, but we could improve this.

When I run this for myself, I see the correct one. This seems to be coming from the GitHub repo and not from anything under the `docs/` directory.

garrettjonesgoogle commented:
Addressed feedback (except one item where I have a question), PTAL

elharo commented on docs/_layouts/default.html:
```diff
@@ -8,8 +8,8 @@
+      {% include jlbp-nav.html %}
       <main role="main" id="page-content">
-        {% include jlbp-nav.html %}
         <section id="jump-content">
```
I need to run the site again, but it looks like there's more navigation inside main, e..g jump-content? Maybe that's the rightand sidebar? Main is for the content only. This is important for accessibility and rendering on non-desktop screens.

garrettjonesgoogle replied:
Done.

garrettjonesgoogle commented:
PTAL, all navigation stuff has been moved outside of `<main>`.

garrettjonesgoogle commented:
Done - Footer updated to black.

garrettjonesgoogle commented:
@elharo what browser are you using to check accessibility?

elharo commented:
That was firefox. Couldn't get Chrome to work.

On Fri, Mar 6, 2020 at 11:09 AM Garrett Jones <notifications@github.com>
wrote:

> @elharo <https://github.com/elharo> what browser are you using to check
> accessibility?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1253?email_source=notifications&email_token=AAHVP2HCH47EPN7XARR3IILRGEN27A5CNFSM4K4MQ3YKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOB4RFQ#issuecomment-595839126>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHVP2F2FSPXAUDP5LTUAVTRGEN27ANCNFSM4K4MQ3YA>
> .
>


-- 
Elliotte Rusty Harold
elharo@macfaq.com

garrettjonesgoogle commented:
- The gray in the left sidebar titles is now fixed
- The redundant title tag in the head tag is now removed

garrettjonesgoogle commented:
I fixed the contrast for highlighted sidebar links.

I'm not sure about the keyboard warnings either. I believe they come from things put there by Glue. It all follows the template provided by Glue at https://glue-docs.appspot.com/docs/components/raw/header#HTML .

elharo commented on docs/_includes/jlbp-footer.html:
```diff
@@ -0,0 +1,8 @@
+<!-- prettier-ignore -->
+<footer>
+  {% if site.google_analytics %}
+  <p>This site uses cookies from Google to deliver its services and to analyze traffic. (<a href='https://policies.google.com/technologies/cookies'>learn more</a>)</p>
```
learn --> Learn

garrettjonesgoogle replied:
Ok

garrettjonesgoogle commented:
> Why do we have both a big View on Github Button and a View Page on Github link?
...
> The link alone is probably sufficient. The primary use case here are readers, not writers and I'd prefer not to put more [chart junk](https://en.wikipedia.org/wiki/Chartjunk) on the page.
> 
> And once that's done, do we really need the top bar at all? Maybe just move the logo into the top of the lefthand nav bar, and then more of the JLBPs fit above the fold.

I removed the "View on Github" button, and I found some more dead css to remove.

We still need the top bar:

1. It is the conventional layout across Google sites to include a header
2. I intend on adding more links in the header, including a search function, I just didn't include them in this PR to simplify it
3. The header provides a "jump to content" accessibility link (part of Glue recommendations)
4. Without the header, mobile has no way to get to the home index. I tried removing the header on desktop only, but Glue seems to depend on header content being available for both desktop/mobile for it to work on mobile at all. 
5. Stripping the header from this PR but then adding it in a following PR creates a bit more unnecessary work

elharo commented on docs/_includes/jlbp-header.html:
```diff
@@ -0,0 +1,61 @@
+      <!-- LOCK up-->
+      <div class="glue-header__container">
+        <div class="glue-header__lock-up">
+          <div class="glue-header__hamburger glue-header__hamburger--first-tier">
```
This doesn't seem to show up and it issues an accessibility warning:
"Clickable elements must be focusable and should have interactive semantics. Learn more"

Can we delete it?

garrettjonesgoogle replied:
This is taken straight from the Glue sample:

https://glue-docs.appspot.com/docs/components/raw/header#HTML

Also it only appears on mobile mode. Shrink the window down to see it appear.

elharo commented on docs/_includes/jlbp-header.html:
```diff
@@ -0,0 +1,61 @@
+        <div class="glue-header__lock-up">
+          <div class="glue-header__hamburger glue-header__hamburger--first-tier">
+            <div class="glue-header__hamburger-wrapper">
+              <button type="button" class="glue-header__drawer-toggle-btn"
```
"Warning

Focusable elements should have interactive semantics."

garrettjonesgoogle replied:
This is taken straight from the Glue sample:

https://glue-docs.appspot.com/docs/components/raw/header#HTML

elharo commented on docs/_includes/jlbp-header.html:
```diff
@@ -0,0 +1,61 @@
+              <p class="glue-header__logo--product">JLBPs</p>
+            </a>
+          </div>
+          <a href="#page-content" class="glue-header__skip-content">
```
"Error

Interactive elements must be labeled. Learn more"

garrettjonesgoogle replied:
This is taken straight from the Glue sample: https://glue-docs.appspot.com/docs/components/raw/header#HTML

elharo replied:
I guess I don't trust Glue as much as you do. When Firefox scans aip.dev it finds a number of problems I know are real issues, so I'm inclined to give it the benefit of the doubt on the ones I don't immediately understand. If you like, we can dig into these further, though in general I'm more comfortable building up from small pieces of code I understand than trying to cut out the large pieces I don't.

garrettjonesgoogle replied:
If there are problems with Glue, I think we should get them fixed at the source instead of 1) trying hack fixes on top of Glue or 2) forging our own path. I would argue that fixing problems at the source should not be in the scope of this PR and shouldn't block it, given the value-add nature of the PR. Are you going to set a bar of perfection in order to approve the PR? We are not a front-end team; we are a team trying to help people solve diamond dependency conflicts.

I would guess that the accessibility issues on aip.dev that have been fixed on jlbp.dev are likely local choices and not issues with Glue.
